src/mlpack/methods/reinforcement_learning/environment/maze.hpp

```diff
@@ -0,0 +1,406 @@
+        {    
+            size_t index = arma::randi(arma::distr_param(0, 3));
+            arma::ivec nextCell = arma::ivec({lastCell(0) + moves[index](0), {lastCell(1) + moves[index](1)}});
+            
+            // Check if particular row and column index are out of bounds of maze.
+            bool outOfBounds = nextCell(0)>= maze.n_rows || nextCell(1)>= maze.n_cols
+        || nextCell(0) < 0 || nextCell(1) < 0 ;
+            if (!outOfBounds && !visited(nextCell(0),nextCell(1))) 
+            {
+                cellsInPath.push(lastCell);
+                cellsInPath.push(nextCell);
+                break;
+            }
+        }
+    }
+
+    // Pop last ({row,column}) from the stack
+      arma::ivec goalCell = cellsInPath.top();
+      cellsInPath.pop();
+
+    // Set the ({row,column}) asa goal cell
+      maze(goalCell(0),goalCell(1)) = 1;
+
+    // Set all other cells in stack to 0 
+    while (!cellsInPath.empty())
+    {
+        arma::ivec lastCell = cellsInPath.top();
+        cellsInPath.pop();
+      
+        maze(lastCell(0),lastCell(1)) = 0;
+    }
+
+    // Randomly set other cells to -1(wall) or 0(path)
+    for (size_t row = 0; row < rows; ++row)
+    {
+      for (size_t col = 0; col < columns; ++col)
+      {
+          if(maze(row,col) !=0 && maze(row,col) !=1 )
+          {
+            maze(row,col) = arma::randi(arma::distr_param(-1, 0));
+          }
+      }
+    }
+
+    // Store goal cell
+    goal = arma::vec({double(goalCell(0)),double(goalCell(1))});
+
+    // Store all the path cells as potential starting points
+    for (size_t row = 0; row < rows; ++row)
+    {
+      for (size_t col = 0; col < columns; ++col)
+      {
+          if(maze(row,col) == 0)
+          {
+            startingPoints.push_back(arma::vec({double(row),double(col)}));
+          }
+      }
+    }
+
+  }
+
+  /**
+   * Get reward and next state based on current
+   * state and current action.
+   *
+   * @param state The current state.
+   * @param action The current action.
+   * @param nextState The next state.
+   * @param transitionGoal The goal for transition
+   * @return reward, it's always 1.0.
+   */
+  double Sample(const State& state,
+                const Action& action,
+                State& nextState,
+                const State& transitionGoal)
+  {
+    // Update the number of steps performed.
+    stepsPerformed++;
+
+    // Make a vector to estimate nextstate.
+    arma::vec currentState {state.Row(), state.Column()};
+    arma::vec direction = directions[action.action];
+
+    arma::vec currentNextState = currentState + direction;
+    nextState.Row() = currentNextState[0];
+    nextState.Column() = currentNextState[1];
+
+    // dont move to that position if invalid
+    bool invalid = false;
+
+    if(currentNextState(0)>= maze.n_rows || currentNextState(1)>= maze.n_cols
+        || currentNextState(0) < 0 || currentNextState(1) < 0 || 
+        maze(currentNextState(0),currentNextState(1))==-1 ){
```

Comment:
Might as well create internal private function for this

Reply:
Sure I can do that.